cmd/nerdctl/cp.go

```diff
@@ -0,0 +1,639 @@
+	}
+	if len(srcSpec.Path) == 0 || len(destSpec.Path) == 0 {
+		return errors.New("filepath can not be empty")
+	}
+
+	client, ctx, cancel, err := newClient(cmd)
+	if err != nil {
+		return err
+	}
+	defer cancel()
+
+	// Supporting only running containers for the moment
+	walker := &containerwalker.ContainerWalker{
+		Client: client,
+		OnFound: func(ctx context.Context, found containerwalker.Found) error {
+			container, err := client.LoadContainer(ctx, found.Container.ID())
+			if err != nil {
+				return err
+			}
+			task, err := container.Task(ctx, cio.Load)
+			if err != nil {
+				return err
+			}
+
+			copyConfig.containerPid = task.Pid()
+
+			status, err := task.Status(ctx)
+			if err != nil {
+				return err
+			}
+
+			switch status.Status {
+			case containerd.Running:
+				if len(srcSpec.Container) != 0 {
+					return copyFromContainer(ctx, cmd, copyConfig)
+				}
+				if len(destSpec.Container) != 0 {
+					return copyToContainer(ctx, cmd, copyConfig)
+				}
+
+				return fmt.Errorf("one of src or dest must be a remote file specification")
+			case containerd.Stopped, containerd.Pausing, containerd.Created, containerd.Paused:
+				return fmt.Errorf("We only support Running Container")
+			default:
+			}
+			return nil
+		},
+	}
+
+	if len(srcSpec.Container) != 0 {
+		copyConfig.container = srcSpec.Container
+	}
+	if len(destSpec.Container) != 0 {
+		copyConfig.container = destSpec.Container
+	}
+
+	n, err := walker.Walk(ctx, copyConfig.container)
+	if err != nil {
+		return err
+	} else if n == 0 {
+		return fmt.Errorf("no such container %s", copyConfig.container)
+	}
+
+	return nil
+}
+
+func splitCpArg(arg string) (fileSpec, error) {
+	i := strings.Index(arg, ":")
+
+	// filespec starting with a semicolon is invalid
+	if i == 0 {
+		return fileSpec{}, errFileSpecDoesntMatchFormat
+	}
+
+	if filepath.IsAbs(arg) {
+		// Explicit local absolute path, e.g., `C:\foo` or `/foo`.
+		return fileSpec{
+			Path: arg,
+		}, nil
+	}
+
+	parts := strings.SplitN(arg, ":", 2)
+
+	if len(parts) == 1 || strings.HasPrefix(parts[0], ".") {
+		// Either there's no `:` in the arg
+		// OR it's an explicit local relative path like `./file:name.txt`.
+		return fileSpec{
+			Path: arg,
+		}, nil
+	}
+
+	return fileSpec{
+		Container: parts[0],
+		Path:      parts[1],
+	}, nil
+}
+
+type fileSpec struct {
+	Container string
+	Path      string
+}
+
+func copyToContainer(ctx context.Context, cmd *cobra.Command, copyConfig cpConfig) (err error) {
+
+	// Init tar Pipe is the first step to process copy to container
+	t := initTarPipe(false, &copyConfig)
+
+	if _, err := os.Stat(t.src); err != nil {
+		return fmt.Errorf("%s doesn't exist in local filesystem", t.src)
+	}
+
+	nerdctlCmd, nerdctlArgs := globalFlags(cmd)
+
+	// TarPipe destination should exist in the container and it should be a directory
+	testDCmd := []string{"test", "-d", t.dst}
+	cmdArgs := append([]string{
+		"exec",
+		t.option.container,
+	}, testDCmd...)
+
```

Comment:
Why do you need to exec? You could just access /proc/PID/root

Reply:
fixed

Reply:
Doesn't seem fixed

Reply:
I am making `test -d /proc/PID/root` isn't enough ?

Reply:
Why do you need to exec `test -d`?

Reply:
@AkihiroSuda https://github.com/containerd/nerdctl/pull/643/files#diff-dc36d8c99566d6fecdc69b09e1781108cb3026a4a5dbbfe31e1df4b4d2572c3bR235

Reply:
Why do you need to exec the `test` command, not just stat?

Reply:
Got it! fixed